sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs

```diff
@@ -90,36 +92,58 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
             return await GetTokenAsync(true, requestContext, cancellationToken).ConfigureAwait(false);
         }
 
-        private async Task<AccessToken> GetTokenAsync(bool isAsync, TokenRequestContext requestContext, CancellationToken cancellationToken)
+        private async ValueTask<AccessToken> GetTokenAsync(bool isAsync, TokenRequestContext requestContext, CancellationToken cancellationToken)
         {
             using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("DefaultAzureCredential.GetToken", requestContext);
 
-            List<Exception> exceptions = new List<Exception>();
+            try
+            {
+                AccessToken token;
+
+                if (_credential != null)
+                {
+                    token = isAsync ? await _credential.GetTokenAsync(requestContext, cancellationToken).ConfigureAwait(false) : _credential.GetToken(requestContext, cancellationToken);
+                }
+                else
+                {
+                    token = isAsync ? await GetTokenFromSourcesAsync(isAsync, requestContext, cancellationToken).ConfigureAwait(false) : GetTokenFromSourcesAsync(isAsync, requestContext, cancellationToken).GetAwaiter().GetResult();
+                }
+
+                return scope.Succeeded(token);
+            }
+            catch (OperationCanceledException e)
+            {
+                scope.Failed(e);
```

Comment:
We're a little inconsistent in places where we do `throw scope.Failed(e)` vs `scope.Failed(e); throw;`?  Is one pattern preferable?

Reply:
`scope.Failed(e)` wraps e in a `AuthenticationFailedException` if it's not already an `AuthencationFailedException`. Possibly bad factoring on my part but it was one common place where I could stick this logic, which was needed in all credential implementations. However after further discussion we decided we didn't want to wrap `OperationCanceledException` as this is an exception that in many cases is initiated by the caller. That is why we have this different pattern for `OperationCanceledException`.

It's possible that if I use the suggestion from @pakrym about moving the rethrow into `scope.Failed` that I could put consolidate this in one place, but I would have to look through all the cases to make sure `OperationCanceledExcpetion` is always handled this way. Thoughts?

Reply:
Actually I take that back. If I were to move the throw into `scope.Failed` I don't think I'd be able to throw without resetting the callstack (since it's not in the context of a catch I would have to `throw e` vs. just `throw`)

Reply:
> Thoughts?

If it's possible the logic is the same everywhere then trying to centralize things makes sense to me, as Pavel says.  FWIW, I pointed out a case today where we may be wrapping `OperationCanceledException` accidentally.